File: frontend/src/lib/invoiceFilename.ts
/**
 * Filename stem aligned with backend ``invoice_filename_stem``: ``TenKH-so``.
 */
export function invoiceFilenameStem(
  customerName: string,
  phone: string | null | undefined,
  fallback: string
): string {
  const rawName = (customerName || "").trim() || fallback;
  const safeName = rawName.replace(/[<>:"/\\|?*\r\n]+/g, "").replace(/\s+/g, "-").slice(0, 120) || fallback;
  const digits = (phone || "").replace(/\D/g, "").slice(0, 20);
  const stem = digits ? `${safeName}-${digits}` : safeName;
  return stem.slice(0, 200);
}

/** Suggested download name for per-order exports (HTML/CSV). */
export function invoiceDownloadFilename(
  customerName: string,
  phone: string | null | undefined,
  ext: string
): string {
  return `${invoiceFilenameStem(customerName, phone, "khach")}.${ext}`;
}

File: backend/app/services/invoice_filename.py
"""Filename stem for per-order exports: ``TenKH-so-dien-thoai``."""

import re
from urllib.parse import quote


def invoice_filename_stem(customer_name: str, phone: str | None, fallback: str) -> str:
    """
    Build a filesystem-safe stem (no extension) from customer name and phone digits.

    Example: ``Nguyễn Văn A`` + ``0909123456`` → ``Nguyễn-Văn-A-0909123456``.
    """
    name = (customer_name or "").strip()
    if not name:
        name = fallback
    tel = "".join(c for c in ((phone or "").strip()) if c.isdigit())
    safe_name = re.sub(r'[<>:"/\\|?*\r\n]+', "", name)
    safe_name = re.sub(r"\s+", "-", safe_name.strip())
    if not safe_name:
        safe_name = fallback
    stem = f"{safe_name}-{tel}" if tel else safe_name
    return stem[:200]


def content_disposition_filename(filename: str) -> str:
    """``Content-Disposition`` with RFC 5987 ``filename*`` for Unicode names."""
    ascii_name = filename.encode("ascii", "replace").decode().replace("?", "_")
    ascii_name = ascii_name.replace('"', "_")[:120]
    if not ascii_name.strip("_"):
        ascii_name = "export"
    return f'attachment; filename="{ascii_name}"; filename*=UTF-8\'\'{quote(filename)}'
